File: problems/字符串/68.文本左右对齐.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<string> fullJustify(vector<string>& words, int maxWidth) {
    vector<string> result;
    int i = 0;
    while (i < words.size()) {
        int j = i + 1, lineLen = words[i].size();
        while (j < words.size() && lineLen + 1 + words[j].size() <= maxWidth)
            lineLen += 1 + words[j++].size();

        int spaces = maxWidth - lineLen, cnt = j - i - 1;
        string line = words[i];
        for (int k = 1; k < j - i; k++) {
            int sp = cnt > 0 ? spaces / cnt : spaces;
            line += string(sp + 1, ' ') + words[i + k];
            spaces -= sp;
            cnt--;
        }
        if (cnt == 0) line += string(maxWidth - line.size(), ' ');
        result.push_back(line);
        i = j;
    }
    return result;
}

int main() {
    // 测试用例
    vector<string> t1 = {"This", "is", "an", "example", "of", "text", "justification."};
    auto r1 = fullJustify(t1, 16);
    for (auto& line : r1) cout << "'" << line << "'" << endl;

    cout << "---" << endl;
    vector<string> t2 = {"What", "must", "be", "acknowledgment", "shall", "be"};
    auto r2 = fullJustify(t2, 16);
    for (auto& line : r2) cout << "'" << line << "'" << endl;
    return 0;
}

File: problems/字符串/68.文本左右对齐.py
#!/usr/bin/env python3
"""
68. 文本左右对齐
"""


def full_justify(words: list[str], max_width: int) -> list[str]:
    result = []
    i = 0
    while i < len(words):
        line_words = [words[i]]
        line_len = len(words[i])
        j = i + 1
        while j < len(words) and line_len + 1 + len(words[j]) <= max_width:
            line_len += 1 + len(words[j])
            line_words.append(words[j])
            j += 1

        if j - i == 1:
            # 只有一个词，左对齐
            result.append(line_words[0] + ' ' * (max_width - len(line_words[0])))
        else:
            # 多个词，左右对齐
            spaces = max_width - line_len + (j - i - 1)
            extra = spaces % (j - i - 1)
            base = spaces // (j - i - 1)
            line = line_words[0]
            for k in range(1, j - i):
                line += ' ' * (base + (1 if k <= extra else 0)) + line_words[k]
            result.append(line)
        i = j
    return result


if __name__ == "__main__":
    # 测试用例
    words = ["This", "is", "an", "example", "of", "text", "justification."]
    for line in full_justify(words, 16):
        print(f"'{line}'")

    print("---")
    words2 = ["What", "must", "be", "acknowledgment", "shall", "be"]
    for line in full_justify(words2, 16):
        print(f"'{line}'")
